operators/src/pro/cache/cache_chunks.rs

```diff
@@ -0,0 +1,398 @@
+            idx: 0,
+        }
+    }
+
+    pub fn element_count(&self) -> usize {
+        self.data.len()
+    }
+}
+
+impl<G: Geometry> Stream for CacheChunkStream<G>
+where
+    FeatureCollection<G>: CacheElementHitCheck,
+    G: ArrowTyped,
+{
+    type Item = Result<FeatureCollection<G>>;
+
+    fn poll_next(
+        mut self: Pin<&mut Self>,
+        _cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Option<Self::Item>> {
+        let CacheChunkStreamProjection { data, query, idx } = self.as_mut().project();
+
+        // return the next tile that is contained in the query, skip all tiles that are not contained
+        for i in *idx..data.len() {
+            let chunk = &data[i];
+            if chunk.cache_element_hit(query) {
+                // TODO: we really should cache the elements bbox somewhere
+                let Ok(chunk) = chunk.filter_cache_element_entries(query) else {
+                    // This should not happen, since we already checked that the element is contained in the query
+                    log::error!("Could not filter cache element entries");
+                    continue;
+                };
+
+                // if the chunk is empty, we can skip it
+                if chunk.is_empty() {
+                    log::trace!("Skipping empty chunk after filtering for query rectangle");
+                    continue;
+                }
+
+                // set the index to the next element
+                *idx = i + 1;
+                return std::task::Poll::Ready(Some(Ok(chunk)));
+            }
+        }
+
+        std::task::Poll::Ready(None)
+    }
+}
+
+pub enum TypedCacheChunkStream {
+    NoGeometry(CacheChunkStream<NoGeometry>),
+    MultiPoint(CacheChunkStream<MultiPoint>),
+    MultiLineString(CacheChunkStream<MultiLineString>),
+    MultiPolygon(CacheChunkStream<MultiPolygon>),
+}
+
+pub trait CacheElementHitCheck {
+    fn cache_element_hit(&self, query_rect: &VectorQueryRectangle) -> bool;
+
+    fn filter_cache_element_entries(
+        &self,
+        query_rect: &VectorQueryRectangle,
+    ) -> Result<Self, CacheError>
+    where
+        Self: Sized;
+}
+
+impl CacheElementHitCheck for FeatureCollection<NoGeometry> {
+    fn cache_element_hit(&self, query_rect: &VectorQueryRectangle) -> bool {
+        let Some(time_bounds) = self.time_bounds() else {
+            return false;
+        };
+
+        time_bounds == query_rect.time_interval || time_bounds.intersects(&query_rect.time_interval)
+    }
+    fn filter_cache_element_entries(
+        &self,
+        query_rect: &VectorQueryRectangle,
+    ) -> Result<Self, CacheError> {
+        let time_filter_bools = self
+            .time_intervals()
+            .iter()
+            .map(|t| t.intersects(&query_rect.time_interval))
+            .collect::<Vec<bool>>();
+        self.filter(time_filter_bools)
+            .map_err(|_err| CacheError::CouldNotFilterResults)
+    }
+}
+
+macro_rules! impl_cache_result_check {
```

Comment:
why do you need a macro? Don't the geometries implement some trait that allows accessing the bounds so you can implement it in a generic way?

Reply:
nope the trait you think of is not implemented for "NoGeometry"